libs/libc/sched/sched_gettid.c

```diff
@@ -0,0 +1,61 @@
+/****************************************************************************
+ * libs/libc/sched/sched_gettid.c
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <unistd.h>
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: nxsched_gettid
+ *
+ * Description:
+ *   Utilise thread local storage to get the thread ID of the currently
+ *   executing thread and cache the result for further queries. Caching this
+ *   result reduces system call loading when using CONFIG_BUILD_KERNEL.
+ *
+ * Input parameters:
+ *   None
+ *
+ * Returned Value:
+ *   On success, returns the thread ID of the calling process.
+ *
+ ****************************************************************************/
+
+#if !defined(CONFIG_BUILD_FLAT) && defined(CONFIG_SCHED_THREAD_LOCAL) && !defined(__KERNEL__)
+
+static __thread pid_t tid;
```

Comment:
could we move to tls_info_s?

Reply:
What's the advantage of using tls_info_s here? I couldn't see how I could integrate it? Possibly I'm just overlooking something.

Reply:
> What's the advantage of using tls_info_s here?
Because all per thread state is held in tls_info_s, and it's bad to use compiler specific key(__thread) in common code.

> I couldn't see how I could integrate it? Possibly I'm just overlooking something.

it's easy, please search tl_errno and see how it work.

Reply:
If we use `tls_info_s`, won't it end up calling `nxsched_get_stackinfo` through `tls_get_info()` and result in a syscall for each gettid anyway?

Reply:
Since you are using kernel mode, syscall can be avoided by enabling CONFIG_TLS_ALIGNED.

Reply:
Ok, yes. I believe I see how it can be used. I've updated my PR.